### PR TITLE
Improve diagnostics for `BaseListExpression`, `StructExpression`, and `AssignmentStatement`

### PR DESCRIPTION
**Pull request by kfcripps**
supersedes https://github.com/p4lang/p4c/pull/3232

### CODE REVIEW

File: testdata/p4_16_errors_outputs/enumCast.p4-stderr [hunk 1]
```diff
@@ -4,13 +4,13 @@ enumCast.p4(26): [--Werror=type-error] error: y: values of type 'int' cannot be
...
              ^
-enumCast.p4(27): [--Werror=type-error] error: AssignmentStatement: values of type 'bit<32>' cannot be implicitly cast to type 'X'
+enumCast.p4(27): [--Werror=type-error] error: y = 32w1: values of type 'bit<32>' cannot be implicitly cast to type 'X'
         y = 32w1; // Error: no implicit cast to enum
           ^
 enumCast.p4(3)
 enum bit<32> X {
              ^
-enumCast.p4(33): [--Werror=type-error] error: AssignmentStatement: values of type 'E2' cannot be implicitly cast to type 'E1'
+enumCast.p4(33): [--Werror=type-error] error: a = b: values of type 'E2' cannot be implicitly cast to type 'E1'
```
Comment (fruffy) [inline testdata/p4_16_errors_outputs/enumCast.p4-stderr:13]: It looks like the assignment statement here is missing quotes, which makes it harder to read.

**Review by fruffy — COMMENTED** [review]
The P4Testgen code seems to fail because of a nullptr access. Unclear why.

File: testdata/p4_16_errors_outputs/enumCast.p4-stderr (hunk 1 above)
Comment (kfcripps) [reply testdata/p4_16_errors_outputs/enumCast.p4-stderr:13]: Do you mean like `"a = b"`?

**Comment by kfcripps** [comment]
> The P4Testgen code seems to fail because of a nullptr access. Unclear why.

I'm going to open separate PRs for each of the IR classes to help isolate the cause.

File: testdata/p4_16_errors_outputs/enumCast.p4-stderr (hunk 1 above)
Comment (fruffy) [reply testdata/p4_16_errors_outputs/enumCast.p4-stderr:13]: Yes

Comment (kfcripps) [reply testdata/p4_16_errors_outputs/enumCast.p4-stderr:13]: Are you requesting to add quotes only to the error messages validated in the tests that this PR affects, or to _all_ error messages that might possibly print an `IR::AssignmentStatement`?

Comment (fruffy) [reply testdata/p4_16_errors_outputs/enumCast.p4-stderr:13]: It is probably prudent to only add quotes to the error messages that are changed in this PR for now. Changing all error messages might be too much too ask for.

**Comment by kfcripps** [comment]
I split this PR into #4357, #4358, and #4360.

**Pull request closed without merging**